### PR TITLE
Test DRSample algorithm

### PR DESCRIPTION
**Pull request by schomatis**
The objective of this PR is to isolate the core of the BucketSample/DRSample algorithm for in-depth testing and audit, to help standardize it, update it in the specification and compare it against other implementations (see [`r2`](https://github.com/nicola/r2/blob/6f27ec762790c42c3b48dee71988e3d6460c8149/src/graph.rs#L47-L64) and [`rust-fil-proofs`](https://github.com/filecoin-project/rust-fil-proofs/blob/3940978de139a1c646018e644f3113be22db218a/storage-proofs/src/drgraph.rs#L199-L216)).

Towards this end one of the main characteristics of the algorithm has been encapsulated in its own structure: the ranges generated for uniform sampling that determine first, the "bucket", a group of consecutive nodes from a logarithmic partitioning, and second, the node itself from that bucket. Why pay attention to the these intermediate sampling ranges and not just the final sampled parent output? The idea is to put aside much of the random components of the algorithm to focus on what we _can_ predict. 

With that in mind, a test was also added with a "fake RNG" that we can set up to "control the randomness" part of the algorithm and make it more predictable. Ideally we should have an independent formula that constructs the ranges and a deterministic sampler (through this fake RNG) to completely layout the behavior of the BucketSample algorithm.

Note: To check that the current implementation wasn't modified inspect commit by commit, particularly the first one [ignoring whitespaces](https://github.com/filecoin-project/drg-attacks/pull/26/commits/321bdba0e0f43c6b239c44a96fe6519beff6c4a8?w=1).

### CODE REVIEW

File: src/graph.rs [hunk 1]
```diff
@@ -80,20 +80,52 @@ pub enum DRGAlgo {
...
+/// Range used for a uniform distribution sample in `Rng::gen_range`: `[low, high)`.
+#[derive(Debug, PartialEq)]
+pub struct UniformSampleRange {
+    pub low: usize,
+    pub high: usize,
+}
+
+/// Ranges used for random sample of a generated parent. The `bucket` range is
+/// deterministic, it only depends on the meta node, while the `node` range is
+/// not, it depends on the bucket samples in the previous range.
```
[the remark this answers is not in the document]
Comment (schomatis) [reply src/graph.rs]: What's the random part of the first range from which we pick the bucket? I think it only depends on the number of nodes.

File: src/graph.rs [hunk 2]
```diff
@@ -837,4 +857,47 @@ pub mod tests {
...
+        let i = v.iter();
+        let mut rng = FakeRNG {
+            iter: &mut i.cycle(),
+        };
+
+        let node = 10;
+        let m = 5;
+
+        let (_, ranges) = Graph::sample_parent_node(node, m, &mut rng);
+        assert_eq!(ranges.bucket, UniformSampleRange { low: 1, high: 7 });
```
[the remark this answers is not in the document]
Comment (schomatis) [reply src/graph.rs]: This is the kind of reasoning I would like to add to the test. The current numbers were just copied from observed output, they weren't predicted.

File: src/graph.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (nikkolasg) [reply src/graph.rs]: Sorry I misunderstood for the number. Yes you're right the first range is picked thanks to the id of the node, and the second range is picked from the number sampled in the first range.

File: src/graph.rs [hunk 3]
```diff
@@ -80,20 +80,52 @@ pub enum DRGAlgo {
     KConnector(usize),
 }
 
+/// Range used for a uniform distribution sample in `Rng::gen_range`: `[low, high)`.
+#[derive(Debug, PartialEq)]
+pub struct UniformSampleRange {
+    pub low: usize,
+    pub high: usize,
+}
```
[the remark this answers is not in the document]
Comment (schomatis) [reply src/graph.rs]: The name encodes not only _what_ it is but also what's its purpose, _how_ do we use it, the same way that an integer is only an integer but can represent different things, even if a type is not comprised of a struct and it's just a single value, it's worth giving it a specific type/name to attach information to it (@laser taught me that :slightly_smiling_face:).

File: src/graph.rs [hunk 4]
```diff
@@ -80,20 +80,52 @@ pub enum DRGAlgo {
...
+pub struct DRSampleRanges {
+    bucket: UniformSampleRange,
+    node: UniformSampleRange,
+}
+
+/// Abstraction over `Rng::gen_range`, which is the only thing we use from
+/// that interface, in order to replace it with a fake RNG for testing purposes.
+/// (We explicitly avoid reusing the `gen_range` name just to avoid multiple
+/// candidates issues.)
+trait SampleRange {
```
[the remark this answers is not in the document]
Comment (schomatis) [reply src/graph.rs]: Yes we can!

**Comment by schomatis** [comment]
> * It seems like it wants to allow reproducible test cases, but it's already the case thanks to the use a given PNRG in the graph's constructor ? i.e. It is already possible to write down the seed and the generated graph in the spec so every implementation could check upon it.

And if the output doesn't match the one in `r2` for example then what?



> * It seems like it wants to verify the ranges are correct, but there are already assertions in the code checking that, and if not enough, maybe a few more assertions in the code would have been sufficient ?

Exactly what assertions do that? What is _the_ correct range? Of course a something like a negative value is incorrect, but there are more subtleties at play here I think (like the `ceil`/`floor` discrepancy with other implementations).

**Comment by schomatis** [comment]
Ok, if we both agree this may be a step in the right direction (although for different motive) I'll fix the name and merge.

**Pull request merged**